io/directory.go

```diff
@@ -394,3 +453,52 @@ func (d *UpgradeableDirectory) AddChild(ctx context.Context, name string, nd ipl
 
 	return nil
 }
+
+// FIXME: Consider implementing RemoveChild to do an eager enumeration if
+//  the HAMT sizeChange goes below a certain point (normally lower than just
+//  zero to not enumerate in *any* occasion the size delta goes transiently
+//  negative).
+//func (d *UpgradeableDirectory) RemoveChild(ctx context.Context, name string) error
+
+// GetNode implements the `Directory` interface. Used in the case where we wrap
+// a HAMTDirectory that might need to be downgraded to a BasicDirectory. The
+// upgrade path is in AddChild; we delay the downgrade until we are forced to
+// commit to a CID (through the returned node) to avoid costly enumeration in
+// the sharding case (that by definition will have potentially many more entries
+// than the BasicDirectory).
+// FIXME: We need to be *very* sure that the returned downgraded BasicDirectory
+//  will be in fact *above* the HAMTShardingSize threshold to avoid churning.
+//  We may even need to use a basic low/high water markings (like a small
+//  percentage above and below the original user-set HAMTShardingSize).
+func (d *UpgradeableDirectory) GetNode() (ipld.Node, error) {
+	hamtDir, ok := d.Directory.(*HAMTDirectory)
+	if !ok {
+		return d.Directory.GetNode() // BasicDirectory
+	}
+
+	if HAMTShardingSize != 0 && hamtDir.sizeChange < 0 {
+		// We have reduced the directory size, check if it didn't go under
+		// the HAMTShardingSize threshold.
+		// FIXME: We should do a partial enumeration instead of the full one
+		//  enumerating until we reach HAMTShardingSize (or we run out of
+		//  entries meaning we need to switch).
+		totalSize := 0
+		ctx := context.Background()
+		// FIXME: We will need a context for this closure to make sure
+		//  we don't stall here. This might mean changing the current
+		//  Directory interface.
+		hamtDir.ForEachLink(ctx, func(l *ipld.Link) error {
+			totalSize += estimatedLinkSize(l.Name, l.Cid)
+			return nil
+		})
```

Comment:
The main options here are:
1) Iterate synchronously in order
2) Iterate asynchronously without strictly caring about the order
3) Iterate asynchronously but trying to stay in order

AFAICT the first two are really easy since we have them via `ForEachLink` and `EnumLinksAsync` the third is a bit trickier (although isn't too far off from `EnumLinksAsync`.

The main issues with synchronous iteration is if a single shard happens to be missing/hard to find that we'll error. The main issue with asynchrnously iterating without caring about the order is that we could potentially end up waiting around for nodes we don't need since we have other ones on disk that could satisfy our query.

Talking with @schomatis it seems like `EnumLinksAsync` shouldn't put us in this situation since it queries mostly in order, however if this becomes an issue we can revisit since this is really just about optimization and doesn't change the data structures at all.

Reply:
Went with number 3 but need to land some of this on the comments to retain the rationale behind the choice.

Reply:
Will be done in https://github.com/ipfs/go-unixfs/issues/105.